runtime/vm/LayoutFFITypeHelpers.hpp

```diff
@@ -399,6 +400,56 @@ class LayoutFFITypeHelpers
 		return primSigType;
 	}
 
+	/* @brief Compare the struct (rather than the union) size and the specified layout ([FF]/[DD]) size
+	 * in bytes to determine whether they are identical on z/OS or not in the case of ALL_SP/ALL_DP.
+	 *
+	 * @param structSig[in] A pointer to the specified struct signature string
+	 * @param structSize[in] the struct size in bytes
+	 * @param layoutSize[in] the layout size in bytes
+	 * @return true if they are identical on z/OS; false otherwise
+	 */
+	static bool
+	isStructWithFFOrDD(const char *structSig, U_32 structSize, U_32 layoutSize)
+	{
+#if defined(J9ZOS390)
+		/* Check whether the specified signature is a union (starting from '{')
+		 * which must be excluded given a union is treated as a non-complex
+		 * type according to the z/OS ABI.
+		 *
+		 * Note:
+		 * [1] The following float/double related structs ('T' stands for float or double) are complex:
+		 * 1) struct {struct {T a[1]; T b[1];}} which is"[1:F1:F]" or "[1:D1:D]" in the signature string.
+		 * 2) struct {struct {T a[1]; T b;} which is "[1:FF]" or "[1:DD]" in the signature string.
+		 * 3) struct {struct {T a; T b[1];} which is "[F1:F]" or "[D1:D]" in the signature string.
+		 * [2] The following float/double related union/structs are non-complex even though they are
+		 * ALL_SP/ALL_DP with 8/16 bytes in size:
+		 * 1) union {T a[2]; T b;} which is "{2:F}" or "{2:D}" in the signature string.
+		 * 2) struct {T a[2];} which is "[2:F]" or "[2:D]" in the signature string.
+		 * 3) struct {struct {T a; T b;} c;} which is "[[FF]]" or "[[DD]]" in the signature string.
+		 */
+		return ('{' != structSig[0]) /* A union starts from '{'. */
+				&& ('[' != structSig[1]) /* A nested struct ("[[FF]]" or "[[DD]]" starts from "[[". */
```

Comment:
this might not be the right fix.  for example, [[[F]]F] ... i.e.  struct { struct { struct { float a;} b; } c; float d;}  this is still a complex i think. i knew we are hammering all kinds of corner cases here.

Reply:
What about `struct { struct { struct { float a; float b;}}}` ? Is it a complex? There might be a bunch of variants to be addressed.

Reply:
no, that is not, since it has only one member.

Reply:
I replaced it with the following code which should cover all these nested cases (with one element).
```
				&& !strstr(structSig, "[[FF]]") /* A struct nested by "[[FF]]" or its variant. */
				&& !strstr(structSig, "[[DD]]") /* A struct nested by "[[DD]]". or its variant. */
```